src/device-manager/java/src/nl/Weave/DataManagement/WdmClientImpl.java

```diff
@@ -0,0 +1,205 @@
+    {
+        init();
+    }
+
+    public void connect(long deviceMgrPtr)
+    {
+        mWdmClientPtr = newWdmClient(deviceMgrPtr);
+    }
+
+    @Override
+    public void close()
+    {
+        if (mTraitMap != null)
+        {
+            Iterator iter = mTraitMap.entrySet().iterator();
+            while (iter.hasNext())
+            {
+                Map.Entry entry = (Map.Entry) iter.next();
+                GenericTraitUpdatableDataSinkImpl traitInstance = (GenericTraitUpdatableDataSinkImpl)entry.getValue();
+                if (traitInstance != null)
+                {
+                    traitInstance.shutdown();
+                }
+            }
+            mTraitMap.clear();
+        }
+
+        if (mWdmClientPtr != 0)
+        {
+            deleteWdmClient(mWdmClientPtr);
+            mWdmClientPtr = 0;
+        }
+
+        mCompHandler = null;
+    }
+
+    @Override
+    public GenericTraitUpdatableDataSink newDataSink(ResourceIdentifier resourceIdentifier, long profileId, long instanceId, String path)
+    {
+        long traitInstancePtr = 0;
+        boolean isExist = false;
+
+        ensureNotClosed();
+
+        if (mTraitMap == null)
+        {
+            mTraitMap = new HashMap<Long, GenericTraitUpdatableDataSink>();
+        }
+
+        traitInstancePtr = newDataSink(mWdmClientPtr, resourceIdentifier, profileId, instanceId, path);
+
+        if (traitInstancePtr == 0)
+        {
+            mCompHandler.onError(new Exception("Fail to create traitInstancePtr."));
+            Log.e(TAG, "traitInstancePtr is not ready");
+            return null;
+        }
+
+        GenericTraitUpdatableDataSink trait = mTraitMap.get(traitInstancePtr);
+        if (trait == null)
+        {
+            trait = new GenericTraitUpdatableDataSinkImpl(traitInstancePtr, this);
+            mTraitMap.put(traitInstancePtr, trait);
+        }
+        else
+        {
+            Log.d(TAG, "trait exists" + profileId);
+        }
+
+        return trait;
+    }
+
+    protected void removeDataSinkRef(long traitInstancePtr)
+    {
+        boolean isExist = false;
+        isExist = mTraitMap.containsKey(traitInstancePtr);
+        if (isExist == true)
+        {
+            mTraitMap.put(traitInstancePtr, null);
+        }
+    }
+
+    @Override
+    public void beginFlushUpdate()
+    {
+        ensureNotClosed();
+        beginFlushUpdate(mWdmClientPtr);
+    }
+
+    @Override
+    public void beginRefreshData()
+    {
+        ensureNotClosed();
+        beginRefreshData(mWdmClientPtr);
+    }
+
+    @Override
+    public CompletionHandler getCompletionHandler()
+    {
+        return mCompHandler;
+    }
+
+    @Override
+    public void setCompletionHandler(CompletionHandler compHandler)
+    {
+        mCompHandler = compHandler;
+    }
+
+    private void onError(Throwable err)
+    {
+        if (mCompHandler == null)
+        {
+            throw new IllegalStateException("unexpected err, mCompHandler is null");
```

Comment:
Could extract a private method here to avoid duplication:

```
private CompletionHandler requireCompletionHandler() {
  return Objects.requireNonNull(
      mCompletionHandler, "No CompletionHandler set. Did you forget to call setCompletionHandler()?");
}
```

Then just use it like:

```
requireCompletionHandler().onError(err);
requireCompletionHandler().onFlushUpdateComplete();
```

etc.

Reply:
Done, thanks